From 3fe8ca57752b942bd43a30fb0b92c96730d1f205 Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Thu, 3 Mar 2016 05:26:42 -0500
Subject: [PATCH] wr-policy: fix avc denials to setrans_t

Reference sources: selinux-policy-3.13.1-128.21.fc22.src.rpm
Changes come from: policy-f22-base.patch

* allow setrans_t use getty_t fd
* allow setrans_t use fd at any level
* allow setrans_t r/w user_tty_device_t chr_file by setting boolean
  'daemons_use_tty' as 'on'

Fix avc denials:

  avc: denied { use } for pid=1779 comm="mcstransd" \
  path="/dev/console" dev="devtmpfs" ino=1038 \
  scontext=system_u:system_r:setrans_t:s15:c0.c1023 \
  tcontext=system_u:system_r:getty_t:s0-s15:c0.c1023 \
  tclass=fd permissive=0

  avc: denied { read write } for pid=1779 comm="mcstransd" \
  path="/dev/console" dev="devtmpfs" ino=1038 \
  scontext=system_u:system_r:setrans_t:s15:c0.c1023 \
  tcontext=root:object_r:user_tty_device_t:s0 \
  tclass=chr_file permissive=0

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/kernel/domain.te  | 13 +++++++++++++
 policy/modules/system/init.te    | 19 +++++++++++++++++++
 policy/modules/system/setrans.te |  1 +
 3 files changed, 33 insertions(+)

--- refpolicy.orig/policy/modules/kernel/domain.te
+++ refpolicy/policy/modules/kernel/domain.te
@@ -13,6 +13,14 @@ policy_module(domain, 1.14.0)
 ## </desc>
 gen_tunable(mmap_low_allowed, false)
 
+## <desc>
+## <p>
+## Allow all domains to use other domains file descriptors
+## </p>
+## </desc>
+#
+gen_tunable(domain_fd_use, true)
+
 # Mark process types as domains
 attribute domain;
 
@@ -184,3 +192,8 @@ allow unconfined_domain_type domain:key
 
 # receive from all domains over labeled networking
 domain_all_recvfrom_all_domains(unconfined_domain_type)
+
+tunable_policy(`domain_fd_use',`
+	# Allow all domains to use fds past to them
+	allow domain domain:fd use;
+')
--- refpolicy.orig/policy/modules/system/init.te
+++ refpolicy/policy/modules/system/init.te
@@ -1134,3 +1134,22 @@ init_get_system_status(init_t)
 
 # MLS trusted for reading from sockets at any level
 mls_socket_read_all_levels(init_t)
+
+## <desc>
+## <p>
+## Allow all daemons the ability to read/write terminals
+## </p>
+## </desc>
+gen_tunable(daemons_use_tty, false)
+
+tunable_policy(`daemons_use_tty',`
+	term_use_unallocated_ttys(daemon)
+	term_use_generic_ptys(daemon)
+	term_use_all_ttys(daemon)
+	term_use_all_ptys(daemon)
+',`
+	term_dontaudit_use_unallocated_ttys(daemon)
+	term_dontaudit_use_generic_ptys(daemon)
+	term_dontaudit_use_all_ttys(daemon)
+	term_dontaudit_use_all_ptys(daemon)
+ ')
--- refpolicy.orig/policy/modules/system/setrans.te
+++ refpolicy/policy/modules/system/setrans.te
@@ -75,6 +75,7 @@ mls_net_receive_all_levels(setrans_t)
 mls_socket_write_all_levels(setrans_t)
 mls_process_read_all_levels(setrans_t)
 mls_socket_read_all_levels(setrans_t)
+mls_fd_use_all_levels(setrans_t)
 
 selinux_compute_access_vector(setrans_t)
 
